src/python/pants/backend/project_info/depgraph.py

```diff
@@ -0,0 +1,200 @@
+# Copyright 2021 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import annotations
+
+import json
+import logging
+from dataclasses import dataclass
+from enum import Enum
+from json import JSONEncoder
+from typing import Iterable, cast
+
+from pants.backend.python.target_types import PythonRequirementsField
+from pants.base.specs import AddressSpecs, MaybeEmptyDescendantAddresses
+from pants.engine.console import Console
+from pants.engine.goal import Goal, GoalSubsystem, LineOriented
+from pants.engine.internals.selectors import MultiGet
+from pants.engine.rules import Get, collect_rules, goal_rule, rule
+from pants.engine.target import Dependencies, DependenciesRequest, Targets
+from pants.util.meta import frozen_after_init
+
+logger = logging.getLogger(__name__)
+
+
+class Granularity(Enum):
+    FILE = "file"
+    TARGET = "target"
+    DIRECTORY = "directory"
```

Comment:
What are you envisioning for these? I'm happy to try to figure out what this means in a [target gen world](https://docs.google.com/document/d/1HpJn2jTWf5sKob6zhe4SqHZ7KWBlX4a6OJOwnuw-IHo/edit world), but don't know your intent.

Reply:
I was thinking that "target" rolls everything up to the explicit BUILD file (usually generator) target level. This may not be useful in a target gen world if we have very few generator targets?   Maybe I should eliminate "target" from this enum.

"directory" is more obvious - we merge all vertices whose files share a directory *and a target*. People are used to thinking in terms of packages, and the package graph is more tractable to look at than the file graph.  This still makes sense in a target gen world. For example, you still see your tests and non-tests in a single directory as two separate vertices.